### PR TITLE
Create ro private /var/tmp dir when /var/tmp is read-only 

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by poettering** [comment]
Hmm, so this only works if if /tmp is writable when /var/tmp is read-only. I am not sure the chance of that being true is that much higher?

Most of our security options are "best effort", i.e. degrade gracefully if some kernel feature or so is not available/prohibited. Maybe we should just make PrivateTmp= work like that too: if we get EROFS creating the private subdir just don't and operate with a non-private /tmp or /var/tmp. This of course means that /tmp and /var/tmp will be as read-only for the service as for the rest of the system, but I think that's totally OK if these dirs are read-only.

And it#s not even a big security issue to gracefully degrade like this: since noone can create files/dirs in the temporary directories nothing can leak or become accessible that shouldn't be accessible...

**Comment by keszybz** [comment]
> this only works if if /tmp is writable when /var/tmp is read-only. I am not sure the chance of that being true is that much higher?

Normally /tmp is tmpfs. Some distros override that, but then I think it's their choice and they lose in this particular case.

> Most of our security options are "best effort", i.e. degrade gracefully if some kernel feature or so is not available/prohibited. Maybe we should just make PrivateTmp= work like that too: if we get EROFS creating the private subdir just don't and operate with a non-private /tmp or /var/tmp. This of course means that /tmp and /var/tmp will be as read-only for the service as for the rest of the system, but I think that's totally OK if these dirs are read-only.

I think there's a difference between downgrading certain features when the system does not support certain options and downgrading because of a runtime condition. The first is predictable and in some way is an implementation choice. But the second can lead to an unexpected leak.

>  Maybe we should just make PrivateTmp= work like that too: if we get EROFS creating the private subdir just don't and operate with a non-private /tmp or /var/tmp. This of course means that /tmp and /var/tmp will be as read-only for the service as for the rest of the system, but I think that's totally OK if these dirs are read-only.

This still creates a possible leak in two ways: first, the ro condition might happen as a result of filesystem error, and then a service that is started later would unexpectedly gain access to file in /tmp. Admittedly low impact even if it occurs. Second, the admin might fix the fs issues and remount rw, and then the service would be left without private /var/tmp in rw mode.

I know that we say that those features are best effort, but in practice people rely on them quite a lot and I think we need to make them robust wrt. to runtime failures.

**Comment by poettering** [comment]
> > this only works if if /tmp is writable when /var/tmp is read-only. I am not sure the chance of that being true is that much higher?
> 
> Normally /tmp is tmpfs. Some distros override that, but then I think it's their choice and they lose in this particular case.

Unfortunately Fedora is one of them though, iirc?

**Comment by poettering** [comment]
> I know that we say that those features are best effort, but in practice people rely on them quite a lot and I think we need to make them robust wrt. to runtime failures.

So here's are two more ideas: when /tmp or /var/tmp are detected read-only then:

1. create a directory /run/systemd/empty/, then ro-bind mount that to the private tmp dir of the service. This way, if /var/tmp and /tmp become writable later on, then you still have to do a restart of the service and until then their /tmp and /var/tmp would stay read-only, cleanly and non-leaking.

2. mount a tiny "emergency" tmpfs instance onto /tmp and /var/tmp if they are read-only. Set a really low limit, i.e. 1M or so, so that really simple things work (such as most bash pipes that are backed by $TMP), but nothing bigger does. When the service goes away the data is released, and memory will not be used without bounds.

I have a slight preference for 1, but 2 would work too, and I think is effectively as good as your current patch but given that we can set much lower limits I like it better I must say

**Comment by keszybz** [comment]
> Unfortunately Fedora is one of them though, iirc?

No. Why do you think that?

> create a directory /run/systemd/empty/, then ro-bind mount that to the private tmp dir of the service. This way, if /var/tmp and /tmp become writable later on, then you still have to do a restart of the service and until then their /tmp and /var/tmp would stay read-only, cleanly and non-leaking.

I guess that this would work. It's mostly the same as my approach, though more complicated. The difference is that things still work if /tmp is read-only. Do you think it's important to support this case too?

**Comment by mrc0mmand** [comment]
> > > this only works if if /tmp is writable when /var/tmp is read-only. I am not sure the chance of that being true is that much higher?
> > 
> > 
> > Normally /tmp is tmpfs. Some distros override that, but then I think it's their choice and they lose in this particular case.
> 
> Unfortunately Fedora is one of them though, iirc?

As @keszybz said it's not Fedora, but RHEL/CentOS does mask the `tmp.mount` unit, to make `/tmp` persistent.

**Comment by poettering** [comment]
> I guess that this would work. It's mostly the same as my approach, though more complicated. The difference is that things still work if /tmp is read-only. Do you think it's important to support this case too?

Well, if it wasn't Fedora that masked tmp.mount, then it was RHEL and Debian. it's probably worth covering them too

**Comment by poettering** [comment]
also needs a rebase

**Comment by keszybz** [comment]
>     1. create a directory /run/systemd/empty/, then ro-bind mount that to the private tmp dir of the service. This way, if /var/tmp and /tmp become writable later on, then you still have to do a restart of the service and until then their /tmp and /var/tmp would stay read-only, cleanly and non-leaking.

I implemented this option. I like it too.

File: src/core/namespace.c [hunk 1]
```diff
@@ -1841,49 +1865,55 @@ static int setup_one_tmp_dir(const char *id, const char *prefix, char **path) {
                 return r;
 
         RUN_WITH_UMASK(0077)
-                if (!mkdtemp(x))
-                        return -errno;
-
-        RUN_WITH_UMASK(0000) {
-                char *y;
+                if (!mkdtemp(x)) {
+                        if (IN_SET(errno, EROFS, ENOSPC))
```
Comment (poettering) [inline src/core/namespace.c]: EDQUOT?

maybe use ERRNO_IS_DISK_SPACE()?

**Comment by keszybz** [comment]
Let's not the green result: now only bionic-i386 is failing with a timeout. I'll repush, and then it'll most likely fail a lot because of Ubuntu woes.

**Comment by keszybz** [comment]
Updated with the one suggested change. I'll merge without waiting for CI since it'll most likely fail anyway.

**Comment by keszybz** [comment]
systemd.systemd passed, let's merge.

**Pull request merged**